CRM/Contribute/BAO/ContributionRecur.php

```diff
@@ -856,13 +859,16 @@ public static function updateOnNewPayment($recurringContributionID, $paymentStat
 
     if (!empty($existing['installments']) && self::isComplete($recurringContributionID, $existing['installments'])) {
       $params['contribution_status_id'] = 'Completed';
+      $params['next_sched_contribution_date'] = '';
     }
     else {
-      // Only update next sched date if it's empty or 'just now' because payment processors may be managing
-      // the scheduled date themselves as core did not previously provide any help.
-      if (empty($existing['next_sched_contribution_date']) || strtotime($existing['next_sched_contribution_date']) ==
-        strtotime($effectiveDate)) {
-        $params['next_sched_contribution_date'] = date('Y-m-d', strtotime('+' . $existing['frequency_interval'] . ' ' . $existing['frequency_unit'], strtotime($effectiveDate)));
+      // Update next_sched_contribution_date in the following circumstances:
+      // - If it's empty.
+      // - If it is older than the contribution receive date.
+      // If we update, set it to the contribution receive date + recur frequency unit * interval
+      if (empty($existing['next_sched_contribution_date'])
+        || strtotime($existing['next_sched_contribution_date']) <= strtotime($contributionReceiveDate)) {
```

Comment:
I'm concerned that this will break the Paypal IPN.  Unless you're intending to merge both this PR plus my PR?

Let's say you don't merge my Paypal IPN PR and this PR is intended to replace that.  This condition will now fail if the following contribution is processed slightly less than the frequency unit.  So for example, let's say for a weekly transaction, on week 5 it's processed on 1-Jan-2020 at 10:00:00.  If the following week, it's processed at 8-Jan-2020 at 09:59:59, this test will fail and the next scheduled contribution date won't update.  At least it will start working again the following week.

Reply:
This stalled on this being too tight - I pulled this & the test into https://github.com/civicrm/civicrm-core/pull/17744 & loosened it up a bit so it should address @andrew-cormick-dockery's issue. I left the time issues out of scope since I suspect they would lead to slippage over time into new days and they seem more like a 'good idea' than an issue being experienced